text/0000-interior.md

```diff
@@ -0,0 +1,525 @@
+trait, and which takes an instance of `Interior<Self>` by value:
+
+```rust
+#[lang="drop_interior"]
+trait DropInterior {
+    fn drop_interior(Interior<Self>);
+}
+```
+
+Define the following functions to convert between `T` and
+`Interior<T>`:
+
+```rust
+impl<T> Interior<T> {
+    unsafe fn from_outer_unsafe(outer: T) -> Self;
+    fn into_outer(this: Self) -> T;
+    fn as_outer(this: &Self) -> &T;
+    fn as_outer_mut(this: &mut Self) -> &mut T;
+    fn of_outer(outer: &T) -> &Self;
+    fn of_outer_mut(outer: &mut T) -> &mut Self;
+}
+```
+
+(Note that `from_outer_unsafe` is made unsafe to avoid a
+backwards-compatibility risk with current `Drop`. This design point
+will also be discussed under **Alternatives**.)
+
+These functions will be used to define conversion traits against
+`Interior<T>`:
+
+```rust
+impl<T> AsRef<T> for Interior<T> { ... }
+impl<T> AsMut<T> for Interior<T> { ... }
+```
+
+(With negative bounds, a safe variant of `from_outer` could be defined
+for !Drop types, and `From<T>` could be implemented for `Interior<T:
+!Drop>`. This is discussed under **Alternatives**.)
+
+With these definitions in place, it is possible to define the legacy
+`Drop` trait in terms of `DropInterior`:
+
+```rust
+impl<T: Drop> DropInterior for T {
+    fn drop_interior(mut this: Interior<T>) {
+        Interior::as_outer_mut(&mut this).drop();
+    }
+}
+```
+
+## Structurally identical
+
+`Interior<T>` is defined to be structurally identical to `T`. There
+are several things we mean by that:
+
+* `Interior<T>` will have the same size and alignment as `T`.
+* The fields of `Interior<T>` will have the same names as the fields of
+`T` and will live at the same memory offsets within the structure.
+* If `T` is an enum, then `Interior<T>` will have the same enum
+discriminant and variants.
+* The accessability of the fields of `Interior<T>` will be identical
+to the accessability of the equivalent fields of `T`.
+
+This definition is intended to support having the conversion from `T`
+to `Interior<T>` (and vice-versa) be a no-op at run-time: the
+introduction of the `Interior<T>` type should introduce *no* run-time
+overhead.
+
+## Drop-glue functions
+
+In this design, we have eliminated the need for the drop-glue function
+for a type to be compiler-generated. Rather, we make the drop-glue
+function (called on scope-based clean-up for a `DropGlue`-value)
+explicit, with an associated lang-item:
+
+```rust
+#[lang="drop_glue_fn"]
+fn drop_glue<T: DropInterior>(arg: T) {
+    // using "unsafe" variant to allow working with `Drop` types (as
+    // opposed to `DropInterior` types that do NOT implement `Drop`).
+    DropInterior::drop_interior(unsafe { Interior::from_outer_unsafe(arg) });
+}
+```
+
+(Note that, under this proposal, all drop-hooks are invoked via
+`DropInterior`, so restricting the `T` argument to implementors of
+`DropInterior` will cover all types that require `DropGlue`.) This
```

Comment:
@eddyb What is place? Why is it needed? Here's how I envision `&in` working:

A function could have an in reference parameter as follows:

    fn consume(val: &in SomeType) { ... }

It would be called analogously to other functions taking references:

    consume(&in foo);

In the caller, the borrow checker treats this the same as moving out of `foo`. Inside `consume` the compiler would implicitly drop `val` at the end of `consume`'s scope or in the event of a panic (unless it is moved or dropped explicitly before then), the same as with an argument passed by value.

For completeness, here is how I envision `&out` operating:

Functions are defined and called analogously to `&in`:

    fn produce(val: &out SomeType) { ... }
    produce(&out foo);

Like with assignment, `foo` must either have no value yet, or be mutable. In the case of a mutable variable with an existing value, the existing value would be dropped before the call. The compiler would ensure that all code paths in `produce` assign a value to `val` (similarly to how in ensures all code paths in a function return a value, today). In the event that a panic occurs during the execution of `produce`, `produce` is responsible for dropping the contents of `val`, if necessary, and the caller assumes `foo` is in an uninitialized state. This is because `produce` knows if `val` has been initialized or not at the point the panic occurred, but the caller does not. Because this means a panic can unwind out of the caller with `foo` uninitialized, `foo` must be legal to leave uninitialized in the event of a panic. In practice, I believe this means it can be a local variable or accessed via an `&in` or `&out` reference, but not an `&mut` reference.